### ISSUE TITLE
Population lifecycle events, limiting the scope of Assetgraph

### ISSUE THREAD

**Issue by Munter** [assetgraph/assetgraph#235]
I have been thinking that our approach to handling preprocessing of assets is not good enough to cover all use cases and isn't extensible enough for people to hook in with their own custom logic.

I had an idea that it might make sense to expose some population lifecycle events, giving the consumer an option to manipulate assets or relations before they are even added to the graph.

This would allow us to think less about all the different kinds of non-web assets that might exist in the graph and how to make them become web assets, thus keeping the scope for Assetgraph narrow.

Obviously we would have to expose an opt-in or opt-out default transformation to make it useful to consumers that don't care for a high degree of customizability. [Accord](https://github.com/jenius/accord) would probably be the best bet here for covering as many compilations as possible and move them out of our range of responsibility.

Obviously the `findOutgoingRelations` methods would still need to be able to pick up the references to these now unknown asset types, but then instead of modeling them as non-web assets just pass the unknown relation to some sort of a resolver-compiler combo.

Does this make sense?

**Comment by slang800** [comment]
That would be great.

There's just a couple things that might need to be considered when designing this:

Compiling an asset could create multiple new assets... like compiling CoffeeScript can create a JS file and a sourcemap. We could deal with that by writing the sourcemap to the disk and then letting Assetgraph find it using the link in the JS file, but that feels pretty hacky. So, I think that the resolver-compiler will need to be able to add multiple Assets when run (rather than just being a straight transformation).

Also, a lot of the compilers (like Jade or Stylus) access files on the disk, relative to their current path. This will pose a problem because with Assetgraph the files could be anywhere (especially in the case of manually added assets). CoffeeScript & LESS are very well encapsulated, but others aren't so nice. I've been looking into making them take a "File" object or something like that to prevent them from accessing the disk directly, but that'll take quite a bit of work.

**Comment by Munter** [comment]
Can sourcemaps be base64 inlined and still be valid? If so, that would solve that problem. If we haven't already added support for that in assetgraph we certainly should.
Writing to disk and reading it back seems like a hacky solution in comparison.

For the precompilers that access assets on disk, assuming we hook into some sort of an `addRelation` event, the complete file url will be available to the compiler if it needs it. However the compiler choses to gather its dependencies would be out of scope for Assetgraph. I've had the same when compiling Sass with `@import` statements. It works fine just letting the compiler figure it out, as long as you don't require special magic with multiple search directories for includes (worst design idea ever).

@mafintosh had a great idea about wrapping the whole precompiling mess into a mock file system. Reading from it would produce the compiled asset. This would be a great abstraction, since everybody is currently reimplementing precompiling over and over and over for development, build and tests. A file system wrapper would be far superior

**Comment by slang800** [comment]
Yep, they can be inlined (and I actually have this working in my fork of accord), but some people might want to have it separate because you can keep the sourcemap in production, as a separate file, and it won't be downloaded by anyone unless their devtools are open. Also the base64'ed version is like 1.5 times the original size, which can be a problem if the sourcemap is big.

And yep -  a mock filesystem is basically what I'm going for with [fobject](https://www.npmjs.com/package/fobject). A way of duck-punching the `fs` object would make for a drop-in replacement, though... So if I can't modify the compiler then I'll probably have to do that.

What do you mean by "reimplementing precompiling over and over"?

**Comment by mafintosh** [comment]
@slang800 what i meant by implementing precompiling as a filesystem i meant as a real filesystem using fuse (like [difffs](https://github.com/mafintosh/difffs))

**Comment by slang800** [comment]
Oh, that makes sense. I never even considered using fuse like that, but that would be a really clever abstraction. Also, the fact that it wouldn't need to start the compilation till the file is accessed would remove a lot of work up-front... might make for some strangely long read-times when a particular file hasn't been cached yet, but other than that it would be really nice.

**Comment by Munter** [comment]
Base64 inlining the source maps is just as a delivery mechanism for assetgraph. It's just a way of handing over two or more assets in the same file. Separating them out again later is something assetgraph is very good at, so I don't see size as a concern there

**Comment by slang800** [comment]
Oh yeah, if assetgraph is going to be separating them out anyway then that's no problem at all.

**Comment by Munter** [comment]
I have started work on a preprocessing file system here: https://github.com/Munter/fusile

So far it's a complete ripoff of other peoples code, mostly @mafintosh's. Any actualy file watching and preprocessing is TODO for now
